### ISSUE TITLE
Unlimited user stylesheets

### ISSUE THREAD

**Issue by ersi-dnd** [OtterBrowser/otter-browser#533]
Right now there's an option to apply just one sticky user style sheet. In Opera there was a whole directory full of user stylesheets that could be selected, applied and switched at will with menu, button, keyboard shortcut, and/or mouse gesture. I hope this is planned :)
![opera](https://cloud.githubusercontent.com/assets/6865273/5129540/e973d332-70eb-11e4-9efa-32dc3338a90b.gif)

**Comment by landroni** [comment]
@Emdek 
*Just reminding that I would like to skip author vs user mode completely and make it simpler*

Having used Author/User mode in Opera for a couple of weeks now, I must actually admit that this arrangement makes a whole lot of sense. It's very flexible, and if the defaults are right, it shouldn't confuse users. 

If you go ahead with a simplified solution, please make sure that the flexibility isn't lost, and that it's easy for the users to switch between Site rendering and User-desired rendering...

**Comment by ersi-dnd** [comment]
When there is no User mode and Author mode, how is user style sheet supposed to work? Am I supposed to point UserStyleSheet path to a file and then it works? And when I don't want it anymore, then I should remove the file from the path?

I haven't tried user styles in Otter, because I am waiting for User mode and Author mode or some other option to change the styles on the fly - as many style files as I want, just like in Opera.

**Comment by Emdek** [comment]
@ersi-dnd, yes, you can set path and later unset it, also in Website Preferences.
Multiple stylesheets to apply in fashion similar to classic Opera should be available later.
But I'm not sure if these two modes are achievable.

**Comment by ersi-dnd** [comment]
Right now I put a user.css into Content/UserStyleSheet. The file has this content


> * {;}

The result: Tabbar becomes enlarged. The user.css in that path is supposed to change webpages, not GUI, right?

There is a purpose to stylesheets

**Comment by Emdek** [comment]
@ersi-dnd, it's unrelated bug in Preferences dialog that I'm constantly forgetting to fix.
Although there is also ability to set styleheet affecting UI (Qt widgets support CSS subset). ;-)

**Comment by ersi-dnd** [comment]
There is a purpose to user stylesheets - like @landroni above, I modify webpages to reduce eyestrain which is becoming an issue as I'm getting older. In Opera the styles worked wonderfully, the best. Only text browsers in terminal emulators (Elinks, w3m) are better.

Next best with user styles currently is the Mozilla family. In FF (and Palemoon and Seamonkey) you can 

1. Go to Preferences/Content/Colors 
2. Choose your colours for links 
3. Select "Use system colors" + Always
4. Voila: Webpage colours are overriden by my system style, my link colours are respected, while webpage layout and scripts remain the same. Very smart.

It would be great if Otter would provide the same. Now, I have noticed that this option to apply system style over webpages is not available in the Chrome family, so possibly webkit&webengine cannot do this. In this case, there should be option to do precisely what Opera used to provide - unlimited user stylesheets + option to switch webpage CSS off. Right now I don't see that webpage CSS can be set off in Otter.

**Comment by annulen** [comment]
@ersi-dnd It's possible to override any settings made by author's style sheets with user style sheet, if you add `!important` to properties

**Comment by ersi-dnd** [comment]
I know that. But I want to override everything and start from scratch. 

Opera's User mode disabled webpage CSS.

Mozilla & Co still retain this menu http://1.bp.blogspot.com/_QIwrz7TuJHs/SofDCqfZa6I/AAAAAAAAA3U/kx0MInkI8eE/s400/Web+Page+Garbled+Or+No+Graphics+-+C.PNG (note that when a webpage has several overlapping or alternate styles, this menu shows them all so they can be disabled/enabled individually)

Otter browser has no function yet to easily disable all webpage CSS and to enable a random selection of individual CSS files.

**Comment by ersi-dnd** [comment]
Now I found an amazing extension that does what we need also on Chromeses https://chrome.google.com/webstore/detail/high-contrast/djcfdncoelnlbldjfhinnjlhdjlikmph

Is there a way to install Chrome extensions to Otter?

**Comment by beastie1** [comment]
> Is there a way to install Chrome extensions to Otter?
Not yet, see #42

**Comment by ersi-dnd** [comment]
Luckily I found a CSS that does one thing that I need: Darken the webpage design https://userstyles.org/styles/2187/myfavolours-global-custom-color-scheme

There are other things that old Opera's user CSS does. It would be great to switch between those files, and between the author design, swiftly and conveniently like in old Opera.

**Comment by Emdek** [comment]
@ersi-dnd, if someone managed to do that using **standard** DOM APIs then we can do that too. @annulen already noted yesterday that there is such API that could be useful in this context:
https://developer.mozilla.org/en-US/docs/Web/API/Document/preferredStyleSheetSet (and related)

**Comment by landroni** [comment]
@Emdek 
Somewhat related... One thing that I'm sorely missing is a single action to enable/disable a stylesheet on a given page, à la `Shift + G` on Opera (`Select author mode > Select user mode`). Is this something we can already do in Otter? 


@ersi-dnd 
> I modify webpages to reduce eyestrain which is becoming an issue as I'm getting older.

You may be interested in using something like `redshift` with perhaps `temp-day=4000` and `gamma=0.95:0.85:0.7` (though personal experimentation is in order to achieve optimal effect, and keep in mind that the optimal settings will evolve with time). I've been using this for years now and I feel that gamma notching and appropriate temp calibration of monitor colors has a disproportionately more useful effect on relieving eye strain than mere dark-ish schemes. In conjunction both are of course useful.

**Comment by Emdek** [comment]
@landroni, it might be more or less doable using JS calls, since there is no C++ API for that.
But as long as it will work then it's not a problem, alternative stylesheets selector already uses it anyway.
I have some other tasks to be done first but hopefully I'll have enough time to take care of it before end of this month, but someone else could give it a try first. ;-)

**Comment by Emdek** [comment]
It's time to revisit this ticket. ;-)
I'm considering adding this to the mix:
https://github.com/openstyles/stylus/wiki/Writing-styles
The question is, how many types of user stylesheets should we support, to not make it too complex for the user. It needs to b decided first, to have stable directory structure for storage (similar to that of classic Opera, with one subfolder for styles specific to given domain and another one for styles applied using menus).
Also we should decide the order in which they should be combined when applying to webpage.

**Comment by Frenzie** [comment]
The Opera method strikes me as essentially fine? Stylish is the most popular these days I suppose.

**Comment by Emdek** [comment]
@Frenzie, after looking up what the heck Stylus does internally (not just hard to parse `@-moz-document` rules, but also variables and conditions) it seems way too complex to include.
So I guess that going for simpler solution from classic Opera would make more sense, since it's way easier to implement and should be easy to use too.
Any idea if there is some documentation somewhere? I would like to extend it a bit (like adding support for `*` for subdomain matching and local `@import` too), but perhaps such rules were already defined (I only have three simple stylesheets around).

I wonder why nobody followed UserJS format for handling URL matching, that would be much easier to handle...

**Comment by ersi-dnd** [comment]
As a general default, if there are no better ideas, why not simply reimplement what is there in Opera? 

I have gotten the impression that the webkit or webengine whichever we use, has its limitations when user styles are applied. Given those limitations, what are the options?

I see that, for example, Vivaldi has not gone beyond the extremely limited implementation of user style options - I may tick some boxes, but they get lost after restart. Also, the different styles do not play well together. In old Opera it was perfect, but that may have been partly due to the original rendering engine.

**Comment by Frenzie** [comment]
Documentation for Opera? I'd say the best best is that blue-colored wiki, but I don't seem to have the address in my bookmarks and you'd probably need Archive.org to look at it.

**Comment by Emdek** [comment]
@ersi-dnd, nah, it's unrelated to engine (although Blink made it harder to implement them properly...), it's up to the application to feed it proper data.

@Frenzie, I see... What about these two ideas then?
`*` is self explanatory, it would work just like matching domains for option overrides.
And `@import` would allow to include other files, using absolute or relative paths (relative to location of parent file).

**Comment by Frenzie** [comment]
I'm not really sure what Opera domain matching you're talking about btw. Did you just mean the site preferences?

**Comment by ersi-dnd** [comment]
Operawiki.org or whatever it was seems to have died. We forgot to download it to save it. Now we must rely on the individual installations of old Opera that we still have.

**Comment by Frenzie** [comment]
I found it thanks to finally remembering "nontroppo" coupled with https://wikiindex.org/Opera_Browser_Wiki

https://web.archive.org/web/20150622025920/http://OperaWiki.info/Opera

That's the best reference to Opera customization I'm aware of, outside of the less structured information contained in the forums.

**Comment by Emdek** [comment]
@Frenzie, I meant our system, used by `override.ini`, so yeah, Website Preferences.

I've found this:
https://web.archive.org/web/20150316135435/http://operawiki.info/OperaUserCSS
But it seems to completely ignore `webpages/` usage...

**Comment by Frenzie** [comment]
@Emdek What are you looking for precisely? The `Applying CSS to specific sites using 'My stylesheet'` section describes what you would do in Opera (and Otter).

**Comment by Emdek** [comment]
@Frenzie, well, I'm referring to method where you place CSS files to `~/.opera/styles/webpages/`, named after domains, so `github.com.css` applies to github.com etc.

**Comment by Frenzie** [comment]
I've never heard of that. Was that a later addition?

**Comment by Frenzie** [comment]
I mean, Moose didn't mention that and the guy worked at Opera…

**Comment by Emdek** [comment]
@Frenzie, dunno, my oldest file there is from 2011.

**Comment by Emdek** [comment]
Anyway, what do you think about this approach?
I would combine it with `user/` subfolder which would b used for storing stylesheets that could be selected using menus and tree based selector like in case of proxies and UA  (depending on UI context).

I'm not sure what to do about current option to select absolute path, sanest way seems to keep it but make it less prominent (or maybe even hide it completely in Website Preferences).

**Comment by ersi-dnd** [comment]
Well, here is how old Opera does it
![Ekraanipilt_2019-04-22_17-25-38](https://user-images.githubusercontent.com/6865273/56505213-b5363480-6523-11e9-8e00-751efb08a610.png)
This is the website preferences dialog. You can put any css file into that slot from anywhere.

**Comment by Frenzie** [comment]
It sounds more complex than just adding a stylesheet through site preference tbh, but on the flipside one might organize stylesheets like that anyway and then it'd be a convenient lack of actions.

**Comment by ersi-dnd** [comment]
The way I meant this thread, it is about unlimited **global** user stylesheets to the style the entire web. Old Opera maybe had some way of applying many site-specific user stylesheets, but it was not explicitly in the GUI and not in the INI files, so I did not use it.
